### ISSUE TITLE
coda_expression_fuzzer: Direct-leak in integer_constant_new

### ISSUE THREAD

**Issue by schwehr** [stcorp/coda#54]
```
#1 0x5577a044e25f in integer_constant_new /proc/self/cwd/third_party/stcorp_coda/libcoda/coda-expr.c:284:12
--
  | #2 0x5577a044d118 in coda_expression_new /proc/self/cwd/third_party/stcorp_coda/libcoda/coda-expr.c:359:20
  | #3 0x5577a0490720 in coda_expression_parse /proc/self/coda/libcoda/coda-expr-parser.y:464:28
  | #4 0x5577a049376e in coda_expression_from_string /proc/self/coda/libcoda/coda-expr-parser.y:1049:9
  | #5 0x5577a03ecfb9 in LLVMFuzzerTestOneInput /proc/self/cwd/third_party/stcorp_coda/fuzz/coda_expression_fuzzer.cc:23:3
```

from 

```c++
#include <stdarg.h>
#include <stddef.h>
#include <stdint.h>

#include <string>

#include "third_party/absl/cleanup/cleanup.h"
#include "third_party/stcorp_coda/libcoda/coda.h"

int printf_black_hole(const char* fmt, ...) {
  va_list args;
  va_end(args);
  return 0;
}

extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
  coda_init();
  auto done = absl::MakeCleanup([] { coda_done(); });

  const std::string exprstring(reinterpret_cast<const char *>(data), size);

  coda_expression *expr = nullptr;
  coda_expression_from_string(exprstring.c_str(), &expr);

  if (!expr) return 0;

  coda_expression_print(expr, printf_black_hole);
  coda_expression_delete(expr);

  return 0;
}
```

[testcase-5913480749645824.zip](https://github.com/stcorp/coda/files/4527158/testcase-5913480749645824.zip)

**Comment by svniemeijer** [comment]
I am not seeing this when I run:
```
$ valgrind codaeval "4Ÿ"
==16163== Memcheck, a memory error detector
==16163== Copyright (C) 2002-2012, and GNU GPL'd, by Julian Seward et al.
==16163== Using Valgrind-3.8.1 and LibVEX; rerun with -h for copyright info
==16163== Command: codaeval 4Ÿ
==16163== 
4
==16163== 
==16163== HEAP SUMMARY:
==16163==     in use at exit: 0 bytes in 0 blocks
==16163==   total heap usage: 6 allocs, 6 frees, 255 bytes allocated
==16163== 
==16163== All heap blocks were freed -- no leaks are possible
==16163== 
==16163== For counts of detected and suppressed errors, rerun with: -v
==16163== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 8 from 6)
```

Also, you should be calling `coda_done()` (although I am not sure it would matter here).

**Comment by schwehr** [comment]
The MakeCleanup causes the coda_done in the lambda to be run when `done` goes out of scope.

**Comment by svniemeijer** [comment]
I tried to replicate your test using:
```
#include <stdio.h>
#include <stdlib.h>
#include "coda.h"

int main()
{
    coda_expression *expr = NULL;
    const char *exprstring = "4\xd9";

    coda_init();

    if (coda_expression_from_string(exprstring, &expr) != 0)
    {
        printf("ERROR: %s\n", coda_errno_to_string(coda_errno));
        coda_done();
        exit(1);
    }

    coda_expression_print(expr, printf);
    printf("\n");
    coda_expression_delete(expr);

    coda_done();
    return 0;
}
```

but running valgrind on this still gives me:

```
==3851== Memcheck, a memory error detector
==3851== Copyright (C) 2002-2012, and GNU GPL'd, by Julian Seward et al.
==3851== Using Valgrind-3.8.1 and LibVEX; rerun with -h for copyright info
==3851== Command: ./exprtest
==3851== 
4
==3851== 
==3851== HEAP SUMMARY:
==3851==     in use at exit: 0 bytes in 0 blocks
==3851==   total heap usage: 22,211 allocs, 22,211 frees, 4,452,932 bytes allocated
==3851== 
==3851== All heap blocks were freed -- no leaks are possible
==3851== 
==3851== For counts of detected and suppressed errors, rerun with: -v
==3851== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 8 from 6)
```

So I don't see where the fuzzer is seeing the leak.

**Comment by schwehr** [comment]
I only see it reproducing in asan-fuzzer mode.  I will investigate and see if I can nail down what is going on

**Comment by svniemeijer** [comment]
Should be fixed in ea4923858ae6aedc9977b85df86c92a3c36ff397

**Comment by schwehr** [comment]
Verified the fix.
